onnxscript/rewriter/pattern.py

```diff
@@ -23,21 +23,96 @@
 # appropriate).
 
 
-class ConstantPattern:
-    def __init__(self, value: int | str | list) -> None:
+class PythonPattern:
+    def __init__(self, value: int | str | Sequence, name: str | None = None) -> None:
         self._value = value
+        self._name = name
 
     @property
-    def value(self) -> int | str | list:
+    def value(self) -> int | str | Sequence:
         return self._value
 
-    def matches(self, value: int | str | list) -> bool:
+    @property
+    def name(self) -> str | None:
+        return self._name
+
+    def matches(self, value: int | str | Sequence) -> bool:
         return value == self.value
 
-    def to_ir(self, model, bindings=None) -> int | str | list:
+    def to_ir(self, model, bindings=None) -> int | str | Sequence:
```

Comment:
Hmmm ... I hadn't thought of the `to_ir` methods (being impacted here). I have been thinking we should probably get rid of these (which exists only as a sort of a poor-substitute for a trace-mode onnxscript integrated into the pattern-IR). Anyway, that is for later, not an issue here.

Reply:
Curious about how we could remove this as to_ir converts Pattern to IR?

Reply:
I created [issue 1403](https://github.com/microsoft/onnxscript/issues/1403) to explain things a bit more. Currently, we use the same `op` object to build the source pattern, as well as to specify the replacement pattern. That has one advantage, namely that the user doesn't have to bother about the distinction. But, in reality, we are doing different things when we are specifying a source-pattern and when we are building the replacement sub-graph. We can use different "builder" objects for the two, and that would avoid the need to couple both of these.